pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.functions.transforms;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.common.schema.SchemaType;
+
+@RequiredArgsConstructor
+public class FlattenStep implements TransformStep {
+  // TODO: Cache schema to flattened schema
+  // TODO: Microbenchmark the flatten algorithm for performance optimization
+  // TODO: Validate flatten delimiter
+  // TODO: Add integration test
+
+  private static final String DEFAULT_DELIMITER = "_"; // '.' in not valid in AVRO field names
+
+  private final Optional<String> delimiter;
+  private final Optional<String> part;
+
+  @Override
+  public void process(TransformContext transformContext) throws Exception {
+    if (!part.isPresent()) {
+      validateAvro(transformContext.getKeySchema());
+      validateAvro(transformContext.getValueSchema());
+      GenericRecord avroKeyRecord = (GenericRecord) transformContext.getKeyObject();
+      GenericRecord avroValueRecord = (GenericRecord) transformContext.getValueObject();
+      transformContext.setKeyObject(flattenGenericRecord(avroKeyRecord));
+      transformContext.setValueObject(flattenGenericRecord(avroValueRecord));
+      transformContext.setKeyModified(true);
+      transformContext.setValueModified(true);
+    } else if (part.isPresent() && "key".equals(part.get())) {
+      validateAvro(transformContext.getKeySchema());
+      GenericRecord avroKeyRecord = (GenericRecord) transformContext.getKeyObject();
+      transformContext.setKeyObject(flattenGenericRecord(avroKeyRecord));
+      transformContext.setKeyModified(true);
+    } else if (part.isPresent() && "value".equals(part.get())) {
+      validateAvro(transformContext.getValueSchema());
+      GenericRecord avroValueRecord =
+          (transformContext.getValueObject()
```

Comment:
@cbornet when running from tests - if I pass a KV schema, I get the value as org.apache.avro.generic.GenericRecord, but if I pass a value only schema (key is a string), I get org.apache.pulsar.client.impl.schema.generic.GenericAvroRecord and I have to consult with the NativeObject.

I would expect both scenarios to give the same result (and that same result to be a GenericAvroRecrod). I'm not sure if this is a side effect [how the test triggers process ](https://github.com/datastax/pulsar-transformations/blob/c217ea692fcd2d67ed0b09bc34a8d0c8a9666def/pulsar-transformations/src/test/java/com/datastax/oss/pulsar/functions/transforms/Utils.java#L68). 

WDYT?

Reply:
That's because of how the function framework works with GenericObject.
If you have a simple Schema.AVRO then the pulsar.GenericRecord you get is a GenericAvroRecord and the wrapped nativeObject is the avro.GenericRecord.
If you have a KV, then the pulsar.GenericRecord is a GenericObjectWrapper and the wrapped nativeObject is a KV<pulsar.GenericRecord, pulsar.GenericRecord>. So when initializing the TransformContext, we unwrap again the key and value.
So I think this code is not useful for real usage, the valueObject will always be avro.GenericRecord. Or did you have issues while testing in integration/live ?

Reply:
no you are right. In integ, I'll always get avro.GenericRecord, but with unit testing, I get the described behavior above. Ideally the code path should be the same for unit/integ testing, let me see if it is possible (and then I can remove the if/else part)